### PR TITLE
Issue #968 import river

### LINKED ISSUE #968: Create iMOD5 RIV converter
It needs to:
- Allocate planar grids to layers
- Distribute planar conductances over layers
- River.from_imod5_data() accepts an allocation and a distribution setting
- Split infiltration factor into separate DRN and RIV package. See [this snippet](https://gitlab.com/deltares/imod/imod-python/-/snippets/2182173), not that this doesn't remove cells with zero conductance. The disv_converter [also creates separate packages](https://github.com/Deltares/imod-python/blob/7e948503a48cc4750f6d350a78ac3b4d3590918a/imod/formats/prj/disv_conversion.py#L438). Set drain elevation equal to river stage. (NB: River stage is defined with respect to model reference level, not river bottom elevation)
- The key of the original RIV package can get an appended, ``"-riv"`` and ``"-drn"`` suffix.
- The split into RIV and DRN package should be logged

Blocked by: https://github.com/Deltares/imod-python/issues/628

### PR DESCRIPTION
**Pull request by luitjansl**
Fixes #968 

# Description
Imports riverpackages from IMOD5. The infiltration factor (which does not exist in MF6) is accounted for by creating a drainage package. 

# Checklist

- [x] Links to correct issue
- [ ] Update changelog, if changes affect users
- [x] PR title starts with ``Issue #nr``, e.g. ``Issue #737``
- [x] Unit tests were added
- [ ] **If feature added**: Added/extended example

### CODE REVIEW

File: imod/mf6/riv.py [hunk 1]
```diff
@@ -162,3 +180,225 @@ def _validate(self, schemata, **kwargs):
...
+        # update the conductance of the river package to account for the infiltration
+        # factor
+        drain_conductance, river_conductance = cls.split_conductance(
+            regridded_package_data["conductance"], infiltration_factor
+        )
+        regridded_package_data["conductance"] = river_conductance
+        regridded_package_data.pop("infiltration_factor")
+
+        river_package = River(**regridded_package_data)
+        # create a drtainage package with the conductance we comptuted from the infiltration factor
```
Comment (JoerivanEngelen) [inline imod/mf6/riv.py:309]:
```suggestion
        # create a drainage package with the conductance we computed from the infiltration factor
```

File: imod/mf6/riv.py [hunk 2]
```diff
@@ -162,3 +180,225 @@ def _validate(self, schemata, **kwargs):
...
+            layered_bottom_elevation = enforce_dim_order(layered_bottom_elevation)
+
+            # due to regridding, the layered_bottom_elevation could be smaller than the
+            # bottom, so here we overwrite it with bottom if that's
+            # the case.
+            layered_bottom_elevation = xr.where(
+                target_bottom > layered_bottom_elevation,
+                target_bottom,
+                layered_bottom_elevation,
+            )
```
Comment (JoerivanEngelen) [inline imod/mf6/riv.py:296]: We need to log this, if it happens

File: imod/mf6/riv.py [hunk 3]
```diff
@@ -162,3 +180,225 @@ def _validate(self, schemata, **kwargs):
...
+            raise ValueError("The infiltration factor should not exceed 1")
+
+        drainage_conductance = conductance * (1 - infiltration_factor)
+
+        river_conductance = conductance * infiltration_factor
+
+        # clean up the packages
+        drainage_conductance = xr.where(
+            drainage_conductance > 0, drainage_conductance, np.nan
+        )
```
Comment (JoerivanEngelen) [inline imod/mf6/riv.py:402]: You can shorten this to:

```suggestion
        drainage_conductance = drainage_conductance.where(
            drainage_conductance > 0
        )
```

File: imod/mf6/riv.py [hunk 4]
```diff
@@ -162,3 +180,225 @@ def _validate(self, schemata, **kwargs):
...
+        river_conductance = xr.where(river_conductance > 0, river_conductance, np.nan)
```
Comment (JoerivanEngelen) [inline imod/mf6/riv.py:403]:
```suggestion
        river_conductance = river_conductance.where(river_conductance > 0)
```

File: imod/tests/test_mf6/test_mf6_riv.py [hunk 5]
```diff
@@ -387,3 +391,76 @@ def test_write_concentration_period_data(concentration_fc):
...
+    target_dis = StructuredDiscretization.from_imod5_data(imod5_dataset)
+
+    (riv, drn) = imod.mf6.River.from_imod5_data(
+        "riv-1",
+        imod5_dataset,
+        target_dis,
+        ALLOCATION_OPTION.at_elevation,
+        DISTRIBUTING_OPTION.by_crosscut_thickness,
+        regridder_types=None,
+    )
```
Comment (JoerivanEngelen) [inline imod/tests/test_mf6/test_mf6_riv.py:408]: I like the API we made!

File: imod/mf6/riv.py [hunk 6]
```diff
@@ -162,3 +180,225 @@ def _validate(self, schemata, **kwargs):
...
+
+    @classmethod
+    def create_infiltration_factor_drain(
+        cls,
+        drain_elevation,
+        drain_conductance,
+    ):
+        # create a drainage package from the river package, to account for the infiltration factor.
+        # this factor is optional in imod5, but it does not exist in MF6, so we mimic its effect
+        # with a Drainage boundary.
```
Comment (JoerivanEngelen) [inline imod/mf6/riv.py:337]: Please convert this to a docstring instead of comment.

File: imod/tests/test_mf6/test_mf6_simulation.py [hunk 7]
```diff
@@ -479,9 +479,12 @@ def test_import_from_imod5(imod5_dataset, tmp_path):
...
+        save_head="last", save_budget="last"
+    )
 
     simulation.create_time_discretization(["01-01-2003", "02-01-2003"])
 
-    # write and validate the simulation
-    simulation.write(tmp_path, False, True, False)
+    # write and validate the simulation.
+    simulation.write(tmp_path, binary=False, validate=True)
+    pass
```
Comment (JoerivanEngelen) [inline imod/tests/test_mf6/test_mf6_simulation.py:490]: Please remove the ``pass`` statement. (I guess a relict of the debugging session?)

**Review by JoerivanEngelen — COMMENTED** [review]
Looks good, my main comment is that I don't think the large amount of deepcopying dicts is a good idea, I have some pragmatic suggestion which avoids a lot of copying overhead. Probably also good to investigate why/when copying is required in the first place.

File: imod/mf6/riv.py (hunk 6 above)
Comment (luitjansl) [reply imod/mf6/riv.py:337]: done

File: imod/mf6/model_gwf.py [hunk 8]
```diff
@@ -206,20 +208,31 @@ def from_imod5_data(
         """
         # first import the singleton packages
         # import discretization
-        dis_pkg = StructuredDiscretization.from_imod5_data(imod5_data, regridder_types)
+
+        dis_pkg = StructuredDiscretization.from_imod5_data(
+            deepcopy_imod5_dict(imod5_data), regridder_types
```
[the remark this answers is not in the document]
Comment (luitjansl) [reply imod/mf6/model_gwf.py:213]: done, removed the deepcopy. it was more of a precaution.

**Review by JoerivanEngelen — APPROVED** [review]
Looks good, thanks for implementing!
